### PR TITLE
UCT/UGNI/UDT: Rework to enable async callback support in UDT

### PR DESCRIPTION
**Pull request by MattBBaker**
Rework of UDT layer so that it now does 'the right thing'. This is the bug free version.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2794/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/216/ for details.

**Comment by MattBBaker** [comment]
@yosefe @shamisp Can you two review?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2190/ for details (Mellanox internal link).

File: src/uct/ugni/udt/ugni_udt_ep.h [hunk 1]
```diff
@@ -13,20 +13,23 @@
 #include <ucs/type/class.h>
 #include <uct/ugni/base/ugni_ep.h>
 
-#define UCT_UGNI_UDT_ANY 0
+#define UCT_UGNI_UDT_ANY(x) 0
```
Comment (shamisp) [inline src/uct/ugni/udt/ugni_udt_ep.h]: hmm. If "x" is always ignored, why do we need it ?

File: src/uct/ugni/udt/ugni_udt_iface.c [hunk 2]
```diff
@@ -8,10 +8,12 @@
 #include "ucs/type/class.h"
 #include "uct/base/uct_md.h"
 
+#include <ucs/async/async.h>
 #include <ucs/arch/cpu.h>
 #include <uct/ugni/base/ugni_iface.h>
 #include "ugni_udt_iface.h"
 #include "ugni_udt_ep.h"
+#include <poll.h>
```
Comment (shamisp) [inline src/uct/ugni/udt/ugni_udt_iface.c]: why do we need poll ?

File: src/uct/ugni/udt/ugni_udt_ep.h (hunk 1 above)
Comment (shamisp) [inline src/uct/ugni/udt/ugni_udt_ep.h]: ok. please fix.

**Comment by shamisp** [comment]
@yosefe please take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2819/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/252/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2212/ for details (Mellanox internal link).

File: src/uct/ugni/udt/ugni_udt_iface.c [hunk 3]
```diff
@@ -185,24 +201,108 @@ static ucs_status_t uct_ugni_udt_iface_query(uct_iface_h tl_iface, uct_iface_att
...
         return;
     }
 
-    ugni_rc = GNI_EpPostDataCancel(self->ep_any);
-    if (GNI_RC_SUCCESS != ugni_rc) {
-        ucs_debug("GNI_EpPostDataCancel failed, Error status: %s %d",
-                  gni_err_str[ugni_rc], ugni_rc);
-        return;
-    }
+    pthread_cancel(self->event_thread);
```
Comment (yosefe) [inline src/uct/ugni/udt/ugni_udt_iface.c]: is it really safe to cancel this way?

File: src/uct/ugni/udt/ugni_udt_iface.c [hunk 4]
```diff
@@ -28,125 +30,138 @@ static ucs_config_field_t uct_ugni_udt_iface_config_table[] = {
...
+    gni_return_t ugni_rc;
+    uint64_t id;
+    int dummy;
+
+    pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, &dummy);
+    while(1){
+        pthread_mutex_lock(&iface->device_lock);
+        while(iface->events_ready) {
+            ucs_trace("Waiting on condition");
+            pthread_cond_wait(&iface->device_condition, &iface->device_lock);
```
Comment (yosefe) [inline src/uct/ugni/udt/ugni_udt_iface.c]: why need pthread_cond_wait if you are waiting anyway with GNI_PostdataProbeWaitById()?

File: src/uct/ugni/udt/ugni_udt_iface.c (hunk 3 above)
Comment (shamisp) [inline src/uct/ugni/udt/ugni_udt_iface.c]: good catch - it is not. You have to signal to the event thread that it is ok to shutdown.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/415/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2918/ for details.

**Comment by MattBBaker** [comment]
@yosefe @shamisp In light of today's call, would it be acceptable to fix the cancel issue and then hold off on the wakeup implementation? I would like to get this merged.

**Comment by MattBBaker** [comment]
@yosefe @shamisp 
Updated the UDT PR to now have 100% fewer instances of pthread_cancel.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3076/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/628/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3077/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/629/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2578/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2579/ for details (Mellanox internal link).

File: src/uct/ugni/base/ugni_iface.h [hunk 5]
```diff
@@ -69,4 +70,15 @@ typedef struct uct_ugni_iface_config {
...
+
+#define uct_ugni_enter_async(x) do {\
+    ucs_trace_async("Taking lock on worker %p", (x)->super.worker); \
+    UCS_ASYNC_BLOCK((x)->super.worker->async);                      \
+    }while(0)
+
+#define uct_ugni_leave_async(x) do {\
+    ucs_trace_async("Releasing lock on worker %p",    (x)->super.worker); \
+    UCS_ASYNC_UNBLOCK((x)->super.worker->async);                        \
+    }while(0)
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.h]: space before while

File: src/uct/ugni/base/ugni_iface.h [hunk 6]
```diff
@@ -69,4 +70,15 @@ typedef struct uct_ugni_iface_config {
...
     uct_iface_mpool_config_t mpool;
 } uct_ugni_iface_config_t;
+
+#define uct_ugni_enter_async(x) do {\
+    ucs_trace_async("Taking lock on worker %p", (x)->super.worker); \
+    UCS_ASYNC_BLOCK((x)->super.worker->async);                      \
+    }while(0)
+
+#define uct_ugni_leave_async(x) do {\
+    ucs_trace_async("Releasing lock on worker %p",    (x)->super.worker); \
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.h]: new line

Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.h]: I mean new line before do

File: src/uct/ugni/base/ugni_ep.h [hunk 7]
```diff
@@ -65,7 +65,8 @@ ucs_arbiter_cb_result_t uct_ugni_ep_process_pending(ucs_arbiter_t *arbiter,
                                                     ucs_arbiter_elem_t *elem,
                                                     void *arg);
 
-static inline void uct_ugni_ep_check_flush(uct_ugni_ep_t *ep){
+static inline void uct_ugni_ep_check_flush(uct_ugni_ep_t *ep)
+{
     if(!ep->outstanding && ep->flush_flag && !ep->arb_flush){
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_ep.h]: this is not your change but spaces are missing before and after the if

**Comment by MattBBaker** [comment]
@shamisp Fixed up your styling issues.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3084/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/638/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2590/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Looks good 👍

**Pull request merged**